### PR TITLE
Cache the GradleWrapper validation and any initial remote archive download by uri

### PR DESCRIPTION
**Pull request by shanman190**

### CODE REVIEW

File: rewrite-core/src/main/java/org/openrewrite/remote/RemoteArchive.java [hunk 1]
```diff
@@ -77,16 +80,34 @@ public class RemoteArchive implements Remote {
      */
     List<String> paths;
 
+    private static Map<URI, Path> downloadCache = new HashMap<>();
```
Comment (jkschneider) [inline rewrite-core/src/main/java/org/openrewrite/remote/RemoteArchive.java:83]: Can we instead pass this through `ExecutionContext` much like we do with `MavenPomCache` interface?

Comment (jkschneider) [reply rewrite-core/src/main/java/org/openrewrite/remote/RemoteArchive.java:83]: What I'm imagining is providing a `RemoteArchiveCache` interface whose default implementation is similar to `LocalMavenArtifactCache`, that would be able to find previously downloaded archives at a predictable path on disk and that wouldn't require static state.

Comment (shanman190) [reply rewrite-core/src/main/java/org/openrewrite/remote/RemoteArchive.java:83]: I don't see any issue with that at the moment. I'll get it adjusted to match that implementation.

Comment (shanman190) [reply rewrite-core/src/main/java/org/openrewrite/remote/RemoteArchive.java]: Added the first iteration via https://github.com/openrewrite/rewrite/pull/3378/commits/215d3fae04b14c994953e21b6e65e3db57abd92d.

**Comment by jkschneider** [comment]
Superseded by #3391

**Pull request closed without merging**